### DISCUSSION TITLE
Fulfilling product orders or how do I find the products a user purchased?

### DISCUSSION THREAD

**Discussion by caged** [pay-rails/pay#512] [Q&A]
:wave: Thank you for the work on this. I have a question related to two previous ones (https://github.com/pay-rails/pay/discussions/443 https://github.com/pay-rails/pay/discussions/460), but I'm still not sure they answer my question after kicking around different ideas on how to solve my issue.

In short, how can I tell which products a user has purchased?  

1. I have multiple products called `things`
2. A user can buy a `Thing` (Stripe Product) through the checkout process. They can also come back later and buy another `Thing` through a second checkout process.
3. I want to know what products the user has purchased. (e.g. `user.purchased_things`).

I've likely been staring at this too long, and I suspect I have been, because this seems like something this libray should be able to do. 

The problem I'm having is that I'm unable to reliably get the associations and ids needed through webhooks or model concern extensions, even when using metadata.

### Checkout with metadata

The metadata sent here is only available in the `checkout.session.completed` webhook.

```ruby
class CheckoutsController < ApplicationController
  before_action :authenticate_user!
  before_action :set_course

  def show
    current_user.set_payment_processor(:stripe)
    current_user.payment_processor.customer

    @payment = current_user.payment_processor.checkout(
      mode: 'payment',
      line_items: 'price_123456789',
      metadata: {
        user_id: current_user.id,
        thing_id: @thing.id
      }
    )
  end
end
```

### Fulfillment in webhooks

**stripe.charge.succeeded**

- `metadata` is empty in the `stripe.charge.succeeded` webhook.
- There's no reference to the Stripe Product.
- I know the customer, and that they were charged, but I don't know the `Thing` they bought.

```ruby
class ChargeSucceeded
  def call(event)
    Rails.logger.info(event)
  end
end

```

**stripe.checkout.session.completed**

* `metadata` is available here!
* There's no way to get a reference to a charge. Now I have a way to get references to the `User` and `Thing`, but no way to find what charge is associated with this session.

```ruby
class CheckoutSessionCompleted
  def call(event)
    Rails.logger.info(event)
  end
end
```

### Fullfillment using Extensions/Concerns

- `metadata` is empty here
- no way to get a reference to the `thing_id`

```ruby
module ChargeExtensions
  extend ActiveSupport::Concern

  included do
    after_create :fulfill_order
  end

  def fulfill_order
    Rails.logger.info(metadata)
  end
end
```

Any ideas where I've went astray here?  Should metadata passed at checkout time, be available in the `Charge` object and in `stripe.charge` callbacks?   

Thanks for your time! 🙏🏻

**Comment by caged** [comment]
I was kind of able to achieve this in the `stripe.checkout.session.completed` since I have access to both the user and the thing, so I can assume the user just made a purchase for the thing, however I'm still not sure how to tie that purchase back to a `Charge` object.

**Comment by cjavilla-stripe** [comment]
This is a little tricky because the line_items for the Checkout Session are not included in the webhook payload. Reference to the Price ID and the Product ID for each line item purchased on a Checkout Session is only available through the Checkout Session (with [expand](https://stripe.com/docs/api/expanding_objects)), not the PaymentIntent or Charge. 

Here's how I'd solve:

```rb
checkout_session_id = event.data.object.id

checkout_session = Stripe::Checkout::Session.retrieve({ id: checkout_session_id, expand: ['line_items'] })
checkout_session.line_items.each do |line_item|
  # Do stuff to create Things here:
  puts line_item.price # Price object
  puts line_item.price.product # Product ID (you can change to `expand: ['line_items.data.price.product']` if you want the full product obj)
end
```

**Comment by caged** [comment]
Thanks for the suggestion @cjavilla-stripe.  Using your example I was able to fill out a local `Purchase` model with enough info to make me comfortable that a user purchased a thing.   

That leave's Pay's `Charge` object kind of orphaned out of the box for basic Product purchases.  Subscriptions, on the other hand, have a foreign key and reference in the `charges` table.  Correct me if I'm wrong, but here is my understanding of the out-of-the-box  functionality (no custom webhooks or concerns):

**You can**
* Charge a user for a Stripe Product using a `price_id`.
* Get a users charges `user.charges` or ` user.payment_processor.charges`

**You can't**
* Reference what Product a user was charged for using the `Charge` API or internal APIs.
* Reference who purchased a Product (`thing.charges`) using the `Charge` object or internal APIs.
* Fulfill product orders without implementing your own webhooks and join model.  [These docs](https://github.com/pay-rails/pay/blob/master/docs/stripe/1_overview.md#fulfilling-orders-after-checkout-completed) suggests Pay will create a subscription for you, but similar features seem to be missing for `Product` purchases.

I believe there are two things that would help, individually, or together. 

1. If Pay had a similar `pay_products` table, mirroring `pay_subscriptions` and if there was a reference to that table in `pay_charges`.
2. If Pay passed the metadata sent at checkout time to all callbacks or if there were references to the `product_id` and/or `price_id` in the Charge callbacks.

FWIW, I'm not complaining about missing functionality if it seems that way.  I'm implementing in-app payments for the first time and trying to stay on the golden path if possible.  I would love to be wrong 😄

Thanks for your time!  I'm happy to mark this as answered, if you believe it to be.  🙏🏻
